jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4960,25 +4960,36 @@ public String descriptorString() {
 	public native boolean isSealed();
 
 	/**
-	 * Returns an array of permitted subclasses as ClassDesc objects related to the calling class.
+	 * Returns the permitted subclasses related to the calling sealed class as an array
+	 * of Class objects. If the calling class is not a sealed class, is an array class,
+	 * or is primitive, then this method returns null instead.
+	 * The order of any classes returned in the array is unspecified, and any classes
+	 * that cannot be loaded are not included in the returned array.
 	 * 
-	 * @return array of ClassDesc objects.
-	 * For a class that is not sealed, an empty array is returned.
+	 * @return array of Class objects if permitted subclasses exist or null if not a sealed class.
+	 * 
+	 * @throws SecurityException if access to any of the classes returned in the array is denied
 	 */
 	@CallerSensitive
-	public ClassDesc[] permittedSubclasses() {
+	public Class<?>[] getPermittedSubclasses() {
 		if (!isSealed()) {
-			return new ClassDesc[0];
+			return null;
 		}
 
 		String[] permittedSubclassesNames = permittedSubclassesImpl();
-		ClassDesc[] permittedSubclasses = new ClassDesc[permittedSubclassesNames.length];
+		ArrayList<Class<?>> permittedSubclasses = new ArrayList<>(permittedSubclassesNames.length);
 
 		for (int i = 0; i < permittedSubclassesNames.length; i++) {
 			String subclassName = permittedSubclassesNames[i];
-			permittedSubclasses[i] = ClassDesc.of(permittedSubclassesNames[i]);
+			try {
+				permittedSubclasses.add(Class.forName(subclassName, false, this.classLoader));
```

Comment:
Does the spec say anything about the initialization status of the classes? If not we will need to verify with the RI what the appropriate behaviour is

Reply:
Spec doesn't mention it, and it looks like initialization doesn't happen in `getPermittedSubclasses` in the RI either.

Reply:
We need to do an access check for each class returned. Class.forName does an access check but it doesnt work in this case since it is being called by Class.getPermittedSubclasses (meaning the caller class/classloader will always be the j.l.Class/bootstrapClassloader). Take a look at the forName function, I think we can use `forNameImpl` and call `forNameAccessCheck` using the real caller